### ISSUE TITLE
Limbs don't rotate properly in modded gravity for Optifine+SmartRender users

### ISSUE THREAD

**Issue by EzerArch** [micdoodle8/Galacticraft#1963]
According to this [wiki page](http://wiki.micdoodle8.com/wiki/Compatibility#Optifine), players who use Optifine are required to install SmartRender, PlayerAPI and RenderPlayerAPI to fix the "can't render oxygen gear" bug". While those APIs do fix the rendering problem, another one ensues: **legs and arms don't rotate properly in modded gravity for SmartRender users**, only on survival mode.

The sitting, aiming and sword-blocking animations don't work properly if the player is being affected by custom gravity, that is, if the player is not on ground (midair, sitting or ridding) in one of the GC dimensions. The limbs don't pivot around the X axis unless walking (which is perfect).

![2015-12-13_04 17 59](https://cloud.githubusercontent.com/assets/6743303/11766127/8b5379d4-a15e-11e5-93f7-80bad9de5f8d.png)

Aiming on ground
![2015-12-13_04 58 30](https://cloud.githubusercontent.com/assets/6743303/11765941/84a63994-a156-11e5-9078-ff41189628b6.png) 

 Aiming midair
![2015-12-13_04 58 32](https://cloud.githubusercontent.com/assets/6743303/11765942/84aa2c34-a156-11e5-89f9-4bcaf4d35b45.png)

Sitting/Ridding
![2015-12-13_04 59 10](https://cloud.githubusercontent.com/assets/6743303/11765940/84a60492-a156-11e5-97b8-d1f6fe8fbbb5.png)

**Comment by EzerArch** [comment]
It seems the 406 release fixed this issue (not sure how).

**Comment by radfast** [comment]
Wow ... so the limb positions are all 100% OK in Creative?

Love those corner blocks btw, and the porthole - which mod?

**Comment by EzerArch** [comment]
@radfast:

> Wow ... so the limb positions are all 100% OK in Creative?

Yes.

> Love those corner blocks btw, and the porthole - which mod?

The corner blocks are from Carpenter's Blocks. The porthole you mean is the Air Lock? It's my resource pack.

**Comment by radfast** [comment]
For the porthole I mean the window with the black surround and white and teal facings.  I wanted to make a block like that in GC...

**Comment by radfast** [comment]
If the limb positions are OK in Creative it means our code must be *mostly* working I hope ... this must be an obscure bug in that case!

**Comment by EzerArch** [comment]
> For the porthole I mean the window with the black surround and white and teal facings.

It's from my resource pack too. :)

**Comment by radfast** [comment]
Can it do connected textures?

**Comment by EzerArch** [comment]
No. But since you are asking, not yet. ;)

**Comment by dj3520** [comment]
Reminds me of RailCraft strengthened glass

**Comment by radfast** [comment]
Please try build 423 for Smart Moving.  In my tests, you need to have RenderPlayerAPI installed alongside Smart Moving to see the GC equipment.  It is 90% working - limbs and head rotate correctly.  (But it goes wrong with thermal armor items still: sometimes they even draw the limbs twice.)

(I think there was a time when we didn't need RenderPlayerAPI for Smart Moving and it just worked, but I don't remember how we did it - if anyone can remember which GC version it was, like 250 or something, that will help!)

**Comment by EzerArch** [comment]
Thank you! I'll test it any time.

As to the porthole gfx, I've just sent you a message.

**Comment by radfast** [comment]
Those tests are with Smart Moving + Render Player API?

I'm seeing:
* GC oxygen mask, oxygen tanks and oxygen gear are rendered correctly, the position is good
* a chestplate is rendering weirdly => arms look too "thick", maybe using player skin on arms of armor
* thermal armor is causing "double render" - 4 legs, 4 arms - I think remove the thermal armor and that problem is fixed
* thermal armor is sometimes rendering with wrong texture, I'm seeing flicker between armor texture and player skin

I did not understand the Chain Helmet point?  Are you saying that wearing an Oxygen Mask causes any helmet to go invisible?

In summary, all these problems are **armor related**.  Wear no armor and no thermal armor, and everything is good.

**Comment by EzerArch** [comment]
> I did not understand the Chain Helmet point? Are you saying that wearing an Oxygen Mask causes any helmet to go invisible?

To be precise, the helmet **overlay** goes invisible in the Inventory UI if wearing an Oxygen Mask.

> a chestplate is rendering weirdly => arms look too "thick", maybe using player skin on arms of armor

The same goes to the leggings (thermal and armor), it renders the player skin, so the legs look thicker. The chestplate arms (thermal and armor) also render the player skin. It doesn't happen with the boots and the helm.

**Thermal Padding comparison:**

build 423 (notice the Z-fight? The boots piece renders the thermal padding graphics whereas the leggings render the player skin):
![2016-01-17_02 10 47](https://cloud.githubusercontent.com/assets/6743303/12375915/91d44294-bcbf-11e5-9a47-c3bf6406f0c9.png)

build 411:
![2016-01-17_02 18 33](https://cloud.githubusercontent.com/assets/6743303/12375941/a277309c-bcc0-11e5-8eaa-d26d47cf248f.png)

**Armor comparison:**

build 423:
![2016-01-17_02 37 22](https://cloud.githubusercontent.com/assets/6743303/12375979/48762d2a-bcc3-11e5-8f89-26038112132e.png)


build 411:
![2016-01-17_02 21 04](https://cloud.githubusercontent.com/assets/6743303/12375975/e6b85432-bcc2-11e5-8273-b108413c9ae1.png)


> In summary, all these problems are armor related. Wear no armor and no thermal armor, and everything is good.

This is me using no armor and no thermal armor, sitting and pulling a bow:

![2016-01-17_02 02 25](https://cloud.githubusercontent.com/assets/6743303/12375890/63b0e850-bcbe-11e5-861d-dc6cacb5efb1.png)

I'm using:
SmartMoving-1.7.10-15.6.jar
SmartRender-1.7.10-2.1.jar
PlayerAPI-1.7.10-1.4.jar
RenderPlayerAPI-1.7.10-1.0.jar

I will test this with GC + SM + OF only and see what happens.

**Comment by radfast** [comment]
Thank you, your screenshots + explanations + comparisons with 411 are very helpful indeed :)

Please update to **RenderPlayerAPI.1.7.10-1.4** and see if it makes a difference.  I'm seeing correct limb positioning in my tests.

Just checking: your texture pack colours the thermal leggings orange?

> To be precise, the helmet overlay goes invisible in the Inventory UI if wearing an Oxygen Mask.

This looks like the Oxygen Mask is being drawn on the same pixels (voxels) where that part of the helmet would go.  I don't see this as a problem / solvable ... we can't draw both unless we use some kind of blending, and blending would not look right either. Anyhow, this is lowest priority, I want to try to fix the other problems first.

**Comment by radfast** [comment]
GC build 424 fixes the helmet issue (and maybe also will fix the duplicate limbs - I'm not sure yet).

All / most remaining issues are related to thermal armor, I think - if you do not wear thermal armor, how is it?

**Comment by EzerArch** [comment]
> Please update to RenderPlayerAPI.1.7.10-1.4 and see if it makes a difference. I'm seeing correct limb positioning in my tests.

WAIT A MINUTE?!!!! Is there an update for 1.7.10? I gonna kill Divisor. BRB. (j/k)

Testing now.

**Comment by radfast** [comment]
I am currently working on this.  I have now:
* fixed the duplicate limbs and Z fighting (I think)
* identified the cause of the "thicker arms" / "player skin overwrites armor arms" issue - that will occur with SmartMoving if you wear any chest-worn oxygen item, e.g. tanks or Oxygen Gear, I know exactly why it is happening now, though it may be hard to fix...

**Comment by radfast** [comment]
So you might want to wait until next build, before testing.

I have not yet identified the reason for leakage of the thermal armor color tinting, so that may be tougher to fix, but lower priority.

**Comment by EzerArch** [comment]
Thank you. :)

**Comment by radfast** [comment]
There were multiple issues happening at the same time.  As far as I can see from the code itself, this has never been working properly before now - but maybe there were other things more obviously wrong or that we were focused on instead.

With the following exception, everything covered in this thread is now fixed, for **Smart Moving + Render Player API** latest versions installed, also for **Render Player API** alone.  (If Smart Moving is installed without Render Player API then GC equipment will not be drawn at all, everything GC will be like vanilla).

The exception which is not fixed (with Smart Moving + Render Player API) is:
* colour tinting of thermal armor will also colour tint the GC equipment worn on the same body part (e.g. thermal helmet tints Oxygen Mask, thermal chest tints oxygen tanks).

But, there's a workaround for that: if you wear regular Minecraft armor on the same body part, then the tinting issue is fixed.  (This should mean the issue is easily fixed, but I just can't find it today...)

**Comment by radfast** [comment]
Micro issues:

I've noticed a faint black ghosting at the end of the player's rendered arm, with Smart Moving + Render Player API installed, not sure if GC is even the cause of that.

Also one small visual glitch around pelvis level, caused by Thermal Chest.

**Comment by EzerArch** [comment]
Ouch, you closed the issue, now I feel a bit embarrassed to post. Just tried 425.

Are you sure you tested it in survival? The animations in creative are perfect, but in survival the animations I'm getting are, say, worse than the initial issue I had reported in the beginning, which was just the arms and legs that didn't pivot vertically. Now the armor, the thermal padding and the player model are out of sync.

[GIF recorded in r425](https://cloud.githubusercontent.com/assets/6743303/12410379/ab737e52-be5b-11e5-9956-9c8c27f33a72.gif)

In order:
1. Creative mode: walking and bow animation, OK.
2. Creative mode: sitting, OK.
3. Survival mode: sitting, desync.
4. Survival mode: walking and bow animation, desync.
5. Survival mode: player looks at the camera, note the oxygen mask is rendered over the iron helmet.
6. Survival mode: player removes the iron armor. Is it the "one small visual glitch around pelvis level"?. It's an issue with the thermal padding png. There's a blank gap between the leggings and the chest. It's easy to fix.

[GIF recorded in r411]
(WIP)

> I've noticed a faint black ghosting at the end of the player's rendered arm, with Smart Moving + Render Player API installed, not sure if GC is even the cause of that.

I'm not getting that. o.o'

**Comment by radfast** [comment]
Not annoying.  I like issue reports which are reproducible :)

I will test this again later today.  I thought I covered Survival in my testing when I passed this as OK, I remember dying to skeletons a few times... but I guess not.

The 'visual glitch around pelvis level' is a horizontal black rectangle, small, at the base of the thermal armour chest model seen from the underside - it's not fully transparent in the "empty" parts of the model.  Must be something in the texture.

3 and 4 are the priorities to fix.

The colour tinting I am finding hard to identify the cause, code-wise.

And helmets on top of helmets on top of helmets (e.g. thermal helm, oxygen mask, armour helmet) may just be impossible to solve, everything is in the same position, but I will look at that finally.

**Comment by radfast** [comment]
OK it's not "worse" than before.  The oxygen tanks and oxygen mask are being rendered in the correct positions, and the thermal armor is not being double rendered (and Z fighting) which was happening before.

New / remaining issues which I can see:
* GC's changes to leg positioning - slowing down the movement when bounding in low gravity - are being applied to the player only, not his armor
* for some reason the player's legs are standing when he inside the boat (though his armor sits)

**Comment by radfast** [comment]
OK that didn't take so long.  It's easier now I have a better understanding of how this part of the code works :)

Issues 3, 4 and 5 in the GIFs are all fixed now.

Colour tinting if thermal armor, but no regular armor, is worn remains a minor issue.

**Comment by radfast** [comment]
Build 427, just up, contains the fixes.

**Comment by dj3520** [comment]
Something I've seen for an unfortunate time now. Not completely sure if it's GC's fault, and if not please point me to the correct place.

Testing with PlayerAPI 1.4, RenderPlayerAPI 1.4, SmartRender 2.1, Optifine HD U C1, and [everything listed here](http://net.dj3520.net:8000/minecraft/modlist.btm) except that Tinkers Construct 1.8.8 is installed and GC is v428 (testing server gets updates first, this listing is of my survival server.)

Testing all armor. Everything works unless specified.

Mutant Creatures v1.4.9 Mutant Skeleton Helmet
![image](https://cloud.githubusercontent.com/assets/8284194/12462367/e54cffc8-bf8b-11e5-82f2-02b20487bde5.png)
Head rotation not correct. Faces north. Turning can change this slightly (acts as body would turn) but doesn't go more than what I can guess is 45 degrees. Walking makes it face north again.

Witchery 0.24.1 All armor (and more)
Same as mutant creatures but effects all armor.
Grotesque potion has same effect
Tinting is OK
Villager Guards (wear leather armor) have leggings that move at a different rate than their legs.

Tinkers Construct 1.8.8 Traveller's gear (I'll be removing this mod anyway)
Same as witchery except goggles work as expected.
Leads to some strange looking results considering the gear
![image](https://cloud.githubusercontent.com/assets/8284194/12462652/74d978fa-bf8d-11e5-9c53-855df93f6adf.png)
Gloves and belt also do not rotate with player

Twilight Forest 2.3.7 All armor
Same as witchery
![image](https://cloud.githubusercontent.com/assets/8284194/12462733/deaedeb4-bf8d-11e5-88b9-fb9653c69cf3.png)

Anything else that needs tested (except tinkers that I will be removing soon) can be done by request. I have not tested in low gravity or with thermal padding.

**Comment by Alaberti** [comment]
@dj3520 I've seen that a lot when I'm running smart moving, and nothing I've experienced points to Galacticraft involvement. I'm no expert, but it seems related to armors not hooking into the player render or...something technical like that. Someone more proficient than I can explain it, I'm certain - the main point is that you'll see this behavior with smart moving and modded armors.

**Comment by EzerArch** [comment]
(I'm using an old spare computer, I can't test anything. Radfast at least can get his deserved rest from me... for now :smile:)

To: @dj3520 and @Alaberti 

99.99% not Galacticraft's fault. SmartMoving (SM), so to say, replaces the rigid vanilla player model (a.k.a. biped model) with another of its own, which can be nicely animated as we know, and all items and armors are also modified accordingly to follow the SM biped model.

The problem is, what happens to the other mods that still rely on the old, plain, rigid vanilla biped model?

![1](https://cloud.githubusercontent.com/assets/8284194/12462367/e54cffc8-bf8b-11e5-82f2-02b20487bde5.png)

![2](https://cloud.githubusercontent.com/assets/8284194/12462652/74d978fa-bf8d-11e5-9c53-855df93f6adf.png)

Ouch!

The mod authors themselves have to add SM support in their own mods, as micdoodle8 and radfast (GC) and machinemuse (MPS) and many others did.

**Comment by EzerArch** [comment]
(I'm using an old spare computer, I can't test anything. Radfast at least can get his deserved rest from me... for now :smile:)

To: @dj3520 and @Alaberti 

99.99% not Galacticraft's fault. SmartMoving (SM), so to say, replaces the rigid vanilla player model (a.k.a. biped model) with another of its own, which can be nicely animated as we know, and all items and armors are also modified accordingly to follow the SM biped model.

The problem is, what happens to the other mods that still rely on the old, plain, rigid vanilla biped model?

![1](https://cloud.githubusercontent.com/assets/8284194/12462367/e54cffc8-bf8b-11e5-82f2-02b20487bde5.png)

![2](https://cloud.githubusercontent.com/assets/8284194/12462652/74d978fa-bf8d-11e5-9c53-855df93f6adf.png)

Ouch!

The mod authors themselves have to add SM support in their own mods, as micdoodle8 and radfast (GC) and machinemuse (MPS) and many others did.

**Comment by Alaberti** [comment]
@EzerArch Thanks for clearing that up!

**Comment by dj3520** [comment]
@EzerArch Thank you for this. As someone who has never tried to tackle
modding, I can only use a programming logic to understand what happens
under the square surface. This seems like an easy thing for mod makers to
accomplish, though I suppose they would simply say that their mod is not
compatible. Plus considering the update frequency of some of them, I
personally won't be reporting this to them.

One last thing. It's tiny. The download pages for RenderPlayer and such
list Galacticraft compatible only for 1.4.7 and down.

**Comment by radfast** [comment]
I have informed RenderPlayer about 12 months ago that Galacticraft 3 is compatible with it, guess I must do that again.... !

With issues with other mods' armors of the type you have described @dj3520, please do this:
* test whether the issue occurs **without Galacticraft installed**
* **if the issue only occurs when Galacticraft is installed** (or if Galacticraft makes the issue worse in some way) then please report it to us, with screenshots of it with + without Galacticraft
* ideally isolate it to the single mod, for example GC + Tinker's Construct + rendering mods - that will also help your game to start more quickly
* any issue of this type it is helpful to test with as many combinations as possible of the following: GC alone, GC + PlayerAPI + RenderPlayerAPI, GC + PlayerAPI + RenderPlayerAPI + SmartMoving (incl. SmartRender), GC + PlayerAPI + RenderPlayerAPI + SmartRender but **no** SmartMoving - each of these different combinations does things internally in a different way, in terms of how limb movement and armor works - note that these rendering mods can all be installed clientside only, it does not matter about these mods on the server
* specify which Forge version, which GC version, and whether Optifine is running.

The results I am expecting are:
(1) generally GC should make no difference
(2) there may be rare situations where another mod's armor does not move correctly at the points when GC changes stuff, for example low gravity bounding on the Moon, or player holding a rocket or flag item

In case (2), RenderPlayerAPI would solve the problem 'automagically' if the other mod would support it, and supporting RenderPlayerAPI would also help the other mod generally, so the best solution would be to get the other mod to do that.  If it's an open source mod I may be able to point them to the way to do it, basically the same as what we have done for GC oxygen equipment and thermal armor (which you could say are like custom types of armor).

If that's not possible I may still be able to make a hacky type of fix through Galacticraft, but that would probably take me 1-2 hours per mod which is time I'd rather spend actually adding features to Galacticraft .... I would perhaps be willing to look at it if it's a major mod that millions of people play.  I'm not unattracted to a situation where installing Galacticraft solves compatibility problems between 2 other major mods, like how GC is sometimes used for power conversion.

**Comment by dj3520** [comment]
@radfast 
>specify which Forge version, which GC version, and whether Optifine is running.

Well I already posted that before all those pictures. Sorry for not including a TL;DR though.
>Testing with PlayerAPI 1.4, RenderPlayerAPI 1.4, SmartRender 2.1, Optifine HD U C1, and [everything listed here](http://net.dj3520.net:8000/minecraft/modlist.btm) except that Tinkers Construct 1.8.8 is installed and GC is v428 (testing server gets updates first, this listing is of my survival server.)

As for testing (MagicLauncher's "load this mod" checkbox and test button helps a ton.)
* Ruled out GC by removing it and seeing the same results.
* Removed SmartMoving and SmartRender. Put GC back in. Worked. WHAT? Tested different armors and survival. Beautiful across the board. 
* Turned SmartRender back on. Problems came back. Interesting.
* Optifine off. Persists. Means Optifine isn't to blame. Oh noes.
* Optifine on and remove RenderPlayerAPI. Oxygen gear renders OK. Went to the moon and saw this: 

![image](https://cloud.githubusercontent.com/assets/8284194/12493138/9fe561ac-c051-11e5-845a-6bead548589c.png)

Good thing I have that hat on!
Landed and jumped (ignore messed up skin texture):
![image](https://cloud.githubusercontent.com/assets/8284194/12493165/d404c270-c051-11e5-9847-9b0caed96659.png)
Note: Witchery's armor did not work before. Tested with IC2 armor, which worked both before and now.
* Turn RenderPlayerAPI back on. Pictured issues persist.

Quick gear viewable by others test
- [X] No Optifine
- [ ] Optifine
- [ ] +PlayerAPI
- [X] +RenderPlayerAPI

So in conclusion: GC was not to blame for what I was seeing, nor Optifine. SmartRender was.
GC still has issues with limbs swinging with mods that have issues with SmartRender. Mods that do support it have no issue. Also the player is still rendered during descent, making the head stick out.
When using GC with Optifine in multiplayer, both PlayerAPI and RenderPlayerAPI are needed to see oxygen gear. Smart Render is **not** required. Mods with SmartMoving support will work with GC's animations at this point, while others have issues like what my 2nd picture shows.

**Comment by radfast** [comment]
> When using GC with Optifine in multiplayer, both PlayerAPI and RenderPlayerAPI are needed to see oxygen gear. 

**This** is a problem I will look at fixing.

The rest seems like a question of the other mods not supporting SmartRender - as you say - but it sounds like your other mods **do** support RenderPlayerAPI, or most of them, which is interesting.  Like I already wrote, GC does things differently internally if SmartRender is present - basically GC chooses to do things the SmartRender way, for 100% compatibility with Smart Moving. I'm not going to change that.  I'd prefer GC movements to be in line with the player body movements, even if another mod has armor which can't keep up.

I think you have a straightforward solution, uninstall Smart Moving + Smart Render and enjoy everything working properly.

**Comment by EzerArch** [comment]
With the power of 2-chunk render distance, fast graphics and smooth lighting off: Gentlemen I've made a test. It's glorious!

![2016-01-21_21 42 09](https://cloud.githubusercontent.com/assets/6743303/12498100/33d2750a-c087-11e5-9cf7-7379c92eb3a9.png)

![2016-01-21_21 44 24](https://cloud.githubusercontent.com/assets/6743303/12498142/80cccc48-c087-11e5-888a-95e138fee961.png)

![2016-01-21_22 02 45](https://cloud.githubusercontent.com/assets/6743303/12498480/10992f90-c08a-11e5-841b-b83617089307.png)

The only remaining issues:

1. "player is still rendered during descent, making the head stick out", as reported @dj3520.
2. The player's arms don't pivot (rotate upwards) with tool swing, bow aim, swork block and sword blow. This happens only if the game returns the player is NOT on a solid block, that is, in a space station dimension and when jumping or in water. On the other hand, those animations are perfect if the player is sitting.

(ADDING AnimGIFs)

> One last thing. It's tiny. The download pages for RenderPlayer and such list Galacticraft compatible only for 1.4.7 and down.

[That's why](http://www.minecraftforum.net/forums/mapping-and-modding/minecraft-mods/1274224-smart-moving?page=358#c7286) and I confirmed that Galacitcraft had lost SM support since 1.5 or so. Then I posted last year ([maybe twice?](http://www.minecraftforum.net/forums/mapping-and-modding/minecraft-mods/1277996-player-api?comment=1740)) to tell him that GC was supporting SM and its APIs again. Possibly he missed my posts.

> Good thing I have that hat on!

LOL.

**Comment by dj3520** [comment]
@radfast Well I only installed Smart Moving because of your testing
requests. But yes a large amount of the mods I use seem to work well
without SmartRender. Though I should point out that those that don't result
in armor limbs that don't match when jumping in low gravity. Guessing that
the way GC changes limbs is similar to SmartMoving, and since these mods
don't support it, their armor has issues. I have pulled SmartRender from my
pack since it does more harm than good. This fixes the ugly rotation issue
but *some* have issues with low gravity. So not 100% fixed but much better.
The low gravity issue makes me think of only one thing...a LimbDict!
Kidding, and I suppose PlayerAPI sort if fills that hole.

This should mean helmets are fine though!

**Comment by dj3520** [comment]
@EzerArch Thanks for laughing. If you have Optifine (guessing you would)
I've found that sacrificing animations improves performance greatly.
Optifine provides a lot of options, but you kind of have to play around
with them to get the best frame rate. Worst yet is that it can vary
depending on what computer hardware is used.

**Comment by radfast** [comment]
> 1.   "player is still rendered during descent, making the head stick out", as reported @dj3520.
  2.  The player arms don't pivot properly (rotate upwards) with tool swing, bow aim, swork block and sword blow. This happens only if the game returns the player is NOT on a solid block, that is, in a space station dimension and when jumping or in water. On the other hand, those animations are perfect if the player is sitting.

Thanks for this, and the new GIFs.  I'm on it.

**Comment by dj3520** [comment]
>the mod's usual armor being rotated to face South

The only time this happens on my end is with SmartRender installed

>Players need to report the incompatibility issues to each of the other mods which is affected 

The list of mods incompatible with GC's custom limb swinging should be the same as the ones who face south with SmartRender. However if the limbs are correct but the head isn't (Mutant Creatures) then technically it's compatible with GC and not compatible with SmartRender.
![image](https://cloud.githubusercontent.com/assets/8284194/12517999/d77a8126-c103-11e5-90bd-c2a24c83240e.png)

This would actually work fine without SmartRender since for some reason it makes some armors face South (though in my testing it's North...) GC doesn't mess with helmets as far as I know, so despite the head facing south with SmartRender, this would work without it perfectly.

If someone could perform my testing on a large modpack (mine I selectively chose) and get back to me on, I'd have a much larger list of mods that do and don't act up. Testing is simple.
1. Insure you additionally have Optifine, PlayerAPI and RenderPlayerAPI installed
2. Create a local creative world for testing.
3. Travel to the moon via dimensiontp
4. Put on a full set of armor belonging from one mod.
5. In 3rd person, jump while walking forward.
6. If nothing appears wrong, this mod is compatible. Add the mod to a "compatible" list so that we don't test again.
7. If instead you get something similar to this:
![image](https://cloud.githubusercontent.com/assets/8284194/12518446/acec4c34-c106-11e5-8c28-5acb9666766d.png)
The mod author will need to be notified. Add to "incompatible" list so they can be notified.
8. Change armor to a mod not tried yet. Repeat from step 5 until all mods with armor are tested.


My list as of currently
OK: Vanilla, GC, IC2, Mutant Creatures, Railcraft
NG: Twilight Forest, TCon, Witchery

In the meantime I figure I'll be making a minecraftforum account to let mod authors know.

>if anyone can remember which GC version it was, like 250 or something, that will help!

First mention of Smart Moving in Jenkins is  0020bece681b0e72255c09e05c825055c07272da though it looks as though it became an Optifine issue with #1656 and GC338

>On ground vs. Jumping:

Looks like he's waving his sword for someone to pass.

**Comment by radfast** [comment]
Sounds like a great idea to let mod authors know of the demand for support of RenderPlayerAPI.  Some of the mods you mentioned are on Github, so that might be better.

I see there as being three possible setups (let's ignore Optifine for now, please test only without it):
1.  Galacticraft only
2.  Galacticraft + PlayerAPI + RenderPlayerAPI
3.  Galacticraft + PlayerAPI + RenderPlayerAPI + SmartRender (probably also Smart Moving)

Galacticraft itself now, since the latest updates, basically **works** in all 3 of these possible setups.  (Let's not worry about the 'waving his sword' issue for now, that is minor and I'll fix it, it's probably connected with the freefall system which makes your limbs come together in freefall.)

"Works" means that:
a. the player model
b. vanilla armor
c. GC oxygen equipment and
d. GC thermal armor
all move together and as intended.  This can be tested both on the Overworld (including holding rockets overhead etc) and in low gravity (the Moon is the best place for testing).

For compatibility with other mods, as long as the mod works with *one* of the systems 1, 2, 3 listed above then we're good here.  I'm not aiming for every mod to work with all of 1, 2, and 3, that is totally outside GC's control and not realistically achievable - we have already done everything in our power to help get to it.  It will then be up to modpack curators to pick a system and choose mods which are all compatible under that system

What I'm understanding from your posts, @dj3520, is that there are mods out there which have custom armor / items which don't work with Galacticraft under *any* of the three systems.

My point of view is that that's a problem because it spoils the player experience.  Really the effort should come from the other mod, because the 'correct' way to deal with these issues is to use RenderPlayerAPI (or SmartRender).  The other mods are basically in the same non-compatibility situation as GC2, before we added / added back compatibility.

But as covered already I may decide to make case-by-case exceptions, either helping out the other mod (by coding for them) or finding a hacky solution in Galacticraft.  It's a trade-off using my coding time for this instead of new features in Galacticraft, which are more interesting for me and for the majority of GC players.  But if it's a **major** mod - for example TiC - then it could be worth doing it, otherwise the Galacticraft experience of a lot of players may be being spoiled.

I will assess this on a case-by-case basis.

What would **help** me is not generalised discussion, and certainly not unexplained screenshots (I have no idea what you are showing me in the screenshots), but instead specific information.  Like "the XY armor module in Z mod looks wrong on the Moon" - tested with 1, 2 and 3.  I will then download the other mod and take a look at XY armor module for myself.  My decision will also be affected by how important XY armor module is in that other mod - is it an item which everyone uses, or is it mostly ignored - and how easy in coding terms the issue is to fix when I look at it.

**Comment by dj3520** [comment]
@radfast All that's needed for someone to do a large scale testing of armors from different mods, and to let them know about the incompatibility. It'd be up to the author to update their mod and support it. I'd do this myself but I don't enjoy large packs and would expect to wait a day after it crashes for it to load again.

>It's a trade-off using my coding time for this instead of new features in Galacticraft,

It's at this point I said "oh no." :grinning: I'd enjoy everything looking right, but less time for features is not acceptable in my book. I don't doubt you're reading the suggestions, some of which are great (such as beam separators and my own thread on redstone controlled spin thrusters) I'd rather see that turned into reality than you doing hacky fixes for mods that aren't yours. I think you'd agree.

>I have no idea what you are showing me in the screenshots

My last post was only to show examples of what a mod that doesn't support SmartRender/PlayerAPI. If the mod doesn't support PlayerAPI and relies on the vanilla model, it will also glitch out with GC or any mod that changes player animations. If you'd like to know more about them, I'd need to know what you'd like to know. I think I've added everything that's needed, but I'd like to work with you as much as possible.

**Comment by EzerArch** [comment]
Just out of curiosity, I've installed Botania:

![2016-01-23_00 19 35](https://cloud.githubusercontent.com/assets/6743303/12527656/966552e8-c166-11e5-85fa-54c5f843de56.png)

See? As I said, mods need to have SM support in order to avoid the above. If you see a mod with an armor that is not just plain retexture from vanilla armor, that animation glitch is likely to happen. Hi-poly armors (e.g. MPS, Botania, Mekanism) are the top victims.

**Comment by dj3520** [comment]
@EzerArch is that with SmartRender? On my testing SmartRender causes
unsupported armor to face south. uninstall it and the only time you'll see
a mes up is when a mod changes limb positioning

**Comment by radfast** [comment]
> 2. The player arms don't pivot properly (rotate upwards) with tool swing, bow aim, swork block and sword blow. This happens only if the game returns the player is NOT on a solid block, that is, in a space station dimension and when jumping or in water. On the other hand, those animations are perfect if the player is sitting.

This is now fixed.  I'm closing this again.  The player being rendered during descent is not related to Smart Moving etc and I don't know whether it's even fixable, though we can try making him invisible I guess.

**Comment by EzerArch** [comment]
> and I don't know whether it's even fixable, though we can try making him invisible I guess.

I believe the problem is, the lander is anchored to the player 1 block down than it's supposed to be. Or the player is 1 block up.

For example: when the lander touches the ground (e.g. y=63.000), my feet pos is y=65.750 and my eye pos y=67.370. The lander is 4 blocks tall, that is, ranging from 63.000 to 67.000, my eye pos is higher than the top height of the lander.

![2016-01-24_19 03 24](https://cloud.githubusercontent.com/assets/6743303/12539174/8c565e6e-c2d3-11e5-8067-b8aff7601b54.png)

This how the player is positioned now:
![2016-01-24_19 50 21](https://cloud.githubusercontent.com/assets/6743303/12539183/c5ff8848-c2d3-11e5-8e32-583cb4798947.png)

If you move the player 1 block down, the player is hidden inside the lander:
![2016-01-24_19 50 46](https://cloud.githubusercontent.com/assets/6743303/12539184/c6058b80-c2d3-11e5-967c-8ea5ffb0430b.png)

Just not sure if it's possible.

**Comment by EzerArch** [comment]
> This is now fixed. I'm closing this again.

Is the fix in 430 build? The fix is not present in-game, at least from I'm seeing, and the change log says nothing about it.

**Comment by radfast** [comment]
431 coming shortly.

**Comment by radfast** [comment]
431 now available.  Also adds redstone control to all powered machines.

**Comment by dj3520** [comment]
Redstone control for machines?! You just made me want to rebuild some over
the top stuff I made in creative.

**Comment by dj3520** [comment]
@EzerArch just a quick verification. Is modular powersuits compatible with Smart Moving? I've seen a few places say yes, but I just tested a suit on mars and it has issues. If it is compatible then something else might be going on.

**Comment by EzerArch** [comment]
@dj3520, Yes. I wouldn't say some animations are not perfect, I'd say instead, they act a bit funny.
![mps](https://cloud.githubusercontent.com/assets/6743303/12534124/05bf70da-c232-11e5-9fae-50ee6aa345b3.gif) 

Just watched your video. You need to have all these 4 SM files installed to get MPS working in GC dimensions:
SmartMoving-1.7.10-15.6.jar
SmartRender-1.7.10-2.1.jar
PlayerAPI-1.7.10-1.4.jar
RenderPlayerAPI-1.7.10-1.4.jar

**Comment by dj3520** [comment]
Ah so it works with Galacticraft only with SmartMoving completely installed. My issue is that I've installed PlayerAPI and RenderPlayerAPI only to fix what Optifine has messed up. Installing SmartRender means many armor types face south. So I'm settling with the lesser of the two evils here.

**Comment by radfast** [comment]
> PlayerAPI and RenderPlayerAPI only to fix what Optifine has messed up

What's that?  I thought we had Optifine licked in latest GC versions (though some reports its still not right in SMP).

**Comment by dj3520** [comment]
I know you need PlayerAPI in order to see your own oxygen gear, though it probably fixes more than just that, I probably didn't see exactly what. RenderPlayerAPI makes it so you can see the oxygen gear of others.

So it looks like in @EzerArch's gif MPS's arm swinging is corrected by SmartMoving? Would that mean that GC isn't doing something that SM is?

**Comment by radfast** [comment]
> I know you need PlayerAPI in order to see your own oxygen gear.

I don't think so - I think GC in SSP mode now works fine with Optifine without needing PlayerAPI.
Seeing gear of others with Optifine is still a problem.  Optifine is a real PITA.

**Comment by EzerArch** [comment]
55 comments, that was a quite long ride. I'm not reporting a bug, but a feature.

Hey, you can get translucent oxygen masks if RenderPlayerAPI is installed.

![2016-01-30_22 59 06](https://cloud.githubusercontent.com/assets/6743303/12699518/149d1648-c7a5-11e5-9adb-bc60620ba56e.png)

**Comment by radfast** [comment]
Lol I didn't know about Et Futurum.  That's a very smart idea for a mod, I like it.

I'm gonna call this a feature, and say no action required :)

But it reminds me, we still have thermal armor colors bleeding in some situations with RenderPlayerAPI.  Bother, I thought I was done with this area.

**Comment by EzerArch** [comment]
> I'm gonna call this a feature, and say no action required :)

No action asked, I'm just letting you know about this cool feature.

Full disclosure: I myself had asked @ganymedes01 to add [helmet translucency](http://www.minecraftforum.net/forums/mapping-and-modding/minecraft-mods/2418997-et-futurum-brings-the-future-to-now?comment=194), because my resource pack's space suits looked so odd with hole to show the player's face instead of glass.

**Comment by radfast** [comment]
Your resource pack's space suits are brilliant by the way - not sure if I've said.
